base_import_match/models/base_import.py

```diff
@@ -0,0 +1,220 @@
+        """Autopatch on init."""
+        models = set(
+            self.browse(
+                cr,
+                SUPERUSER_ID,
+                self.search(cr, SUPERUSER_ID, list()))
+            .mapped("model_name"))
+        for model in models:
+            self._load_autopatch(cr, SUPERUSER_ID, model)
+
+    @api.multi
+    @api.depends("model_id", "field_ids")
+    def _compute_name(self):
+        """Automatic self-descriptive name for the setting records."""
+        for s in self:
+            s.name = "{}: {}".format(
+                s.model_id.display_name,
+                " + ".join(s.field_ids.mapped("display_name")))
+
+    @api.model
+    def _load_wrapper(self):
+        """Create a new load patch method."""
+        @api.model
+        def wrapper(self, fields, data):
+            """Try to identify rows by other pseudo-unique keys.
+
+            It searches for rows that have no XMLID specified, and gives them
+            one if any :attr:`~.field_ids` combination is found. With a valid
+            XMLID in place, Odoo will understand that it must *update* the
+            record instead of *creating* a new one.
+            """
+            newdata = list()
+
+            # Mock Odoo to believe the user is importing the ID field
+            if "id" not in fields:
+                fields.append("id")
+
+            # Needed to work with relational fields
+            clean_fields = [
+                models.fix_import_export_id_paths(f)[0] for f in fields]
+
+            # Get usable rules to perform matches
+            usable = self.env["base_import.match"]._usable_for_load(
+                self._name, clean_fields)
+
+            for row in (dict(zip(clean_fields, r)) for r in data):
+                # All rows need an ID
+                if "id" not in row:
+                    row["id"] = u""
+
+                # Skip rows with ID, they do not need all this
+                elif row["id"]:
+                    continue
+
+                # Store records that match a combination
+                match = self
+                for combination in usable:
+                    match |= self.search(
+                        [(field.name, "=", row[field.name])
+                         for field in combination.field_ids])
+
+                    # When a single match is found, stop searching
+                    if len(match) != 1:
+                        break
+
+                # Only one record should have been found
+                try:
+                    match.ensure_one()
+
+                # You hit this because...
+                # a. No match. Odoo must create a new record.
+                # b. Multiple matches. No way to know which is the right
+                #    one, so we let Odoo create a new record or raise
+                #    the corresponding exception.
+                # In any case, we must do nothing.
+                except ValueError:
+                    continue
```

Comment:
Roadmap idea: add a log to write these exceptions to, so that end users can monitor it and decide what to do to each individual record: ignore or force load.